File: 排序算法/快速排序/QuickSort.java
import java.util.*;

class QuickSort {
    public static <T extends Comparable> void sort(
        List<T> array, int start, int end) {
        if (start >= end)
            return;

        T pivot = array.get(start);
        int pos = start;
        for (int ind=start+1; ind<=end; ind++) {
            if (array.get(ind).compareTo(pivot) > 0)
                continue;
            array.set(pos, array.get(ind));
            for (int ind1=ind; ind1>=pos+2; ind1--)
                array.set(ind1, array.get(ind1-1));
            array.set(++pos, pivot);
        }

        sort(array, start, pos-1);
        sort(array, pos+1, end);
    }
    
    public static void main(String[] args) {
        List<Integer> integers = new ArrayList<>();
        integers.add(5);
        integers.add(3);
        integers.add(2);
        integers.add(9);
        integers.add(7);
        integers.add(0);
        integers.add(1);
        sort(integers, 0, integers.size() - 1);
        System.out.println(integers);
    }
}

File: 排序算法/快速排序/quick_sort.go
package main

import (
	"fmt"
)

func quickSort(a []int, start, end int) {
	if start >= end {
		return
	}
	pos := partition(a, start, end)
	quickSort(a, start, pos-1)
	quickSort(a, pos+1, end)
}

func partition(a []int, start, end int) int {
	if start > end {
		panic("start should more than end")
	}

	pivot := a[start]
	index := start // pivot 所在的位置
	for start < end {
		for end > start {
			if a[end] >= pivot {
				end -= 1
			} else {
				a[index], a[end] = a[end], a[index]
				start += 1
				index = end
				break
			}
		}

		for start < end {
			if a[start] <= pivot {
				start += 1
			} else {
				a[start], a[index] = a[index], a[start]
				end -= 1
				index = start
				break
			}
		}
	}
	return index
}

func QuickSort(a []int) {
	if len(a) == 0 {
		return
	}
	quickSort(a, 0, len(a)-1)
}

func main() {
	a := []int{3, 2, 9, -1, 88, 66, 1, 33, -3, 33, 88, 2, 9, 3}
	QuickSort(a)
	fmt.Println(a)
}
